src/mds/CInode.h

```diff
@@ -1050,6 +1050,15 @@ class CInode : public MDSCacheObject, public InodeStoreBase, public Counter<CIno
                            MDSContext *fin);
   static void dump_validation_results(const validated_data& results,
                                       ceph::Formatter *f);
+  bool has_inline_data() {
+    if (is_normal() && is_file()) {
+      auto pin = get_projected_inode();
+      if (pin->inline_data.version != CEPH_INLINE_NONE) {
```

Comment:
Maybe this is incorrect. In case the inlinedata is enabled then for all the new CInodes the default version is `1` here. Which has no inlinedata.

Reply:
are you concerned that the attempt to `write()` to the 0th file object is a no-op when the inline data length is zero ?
do you want to avoid that round trip to the osd ?

Reply:
IMO, we should also skip uninlining the inline data if the `pin->inline_data.version == 1`. Actually there is no inline data for this inode.

Reply:
Or the uninline data work will iterate all the CInodes and do the fake uninline work, even there is no inline data in any file.

Reply:
for `write`: so, now the write takes place conditionally only if inline_data length is a non-zero positive value
for `create`: if I completely remove the `create` op, then uninline ops fail with ENOENT; at least for newly created files

will refresh the PR once I've addressed comments

Reply:
`write` has been made conditional only if inline data length is non-zero

Reply:
In `void MDCache::uninline_data_work(MDRequestRef mdr)` I see after checking the `if (!in->has_inline_data()) {` if the inlindata version is `1` it will continue do a lot more work after that. IMO we can just skip them ?

Reply:
If we are planning on deprecating and eventually decommissioning **inline data** feature, then we'd need to do all the work that needs to be done to move the information to the 0th file object. Inline data version and the xattr does need to be projected and logged to the journal. I don't think we can skip this work even if **inline data length** is zero irrespective of the version ... since inline data may have been written within the inline data length bounds and truncated to zero length many times. If so, then the inline data version may not be 1 but the inline data length may be zero.

@batrick could you take a peek here and suggest appropriately

Reply:
So, I think there is a structural problem in this code: we shouldn't be looking at the projected version at all. If we're uninlining data then clients should not be writing any new inline data, no? I think before any automatic uninlining is done via scrub, the MDS should ensure there are no inline data writes in play. If there are, scrub should not do any uninlining and note a health warning that inline is still in use.

Also, I forget: how are you preventing new inline writes by clients?

Reply:
>  I presumed that acquiring xlocks causes cap revokes and/or commits pending writes, so that eventually acquiring the xlock in the uninline code path fences the client from causing any racy writes.

Sure (but verify with tests!).

> So, I think that once an xlock has been granted to the uninline code path, the has_inline_data() function need not look at the projected inode.

Um, the client could still have dirtied data (like a truncate) which requires looking at the projected inode.

I take back what I said about not looking at the projected data. That won't be practical. We do however need to prevent clients from doing new inline writes.

Reply:
> > So, I think that once an xlock has been granted to the uninline code path, the has_inline_data() function need not look at the projected inode.
> 
> Um, the client could still have dirtied data (like a truncate) which requires looking at the projected inode.

Wouldn't the client block for `write` (`Fw`) caps for the truncate op ?
Or could things turn ugly with an async truncate op ?

@lxbsz could you comment here about the client behavior in this case

Reply:
> > > So, I think that once an xlock has been granted to the uninline code path, the has_inline_data() function need not look at the projected inode.
> > 
> > 
> > Um, the client could still have dirtied data (like a truncate) which requires looking at the projected inode.
> 
> Wouldn't the client block for `write` (`Fw`) caps for the truncate op ? 

The `Fw` is not a must for the size truncating, please see https://github.com/ceph/ceph/blob/main/src/client/Client.cc#L7868-L7895.

If the `Fx` cap is not issued it will send a sync `setattr` request to MDS, or the clients could buffer the size change.

> Or could things turn ugly with an async truncate op ?
> 

Then in the MDS side, the `Server::handle_client_setattr()` will require the `xlock` for the file. So the truncate request should be queued to wait the scrub uninline work to finish.

I think this should be okay and I don't see a racy for this.

> @lxbsz could you comment here about the client behavior in this case

But to be safe, I suggest to make the current file to be readonly while uninlining the inline data.

> > So, I think that once an xlock has been granted to the uninline code path, the has_inline_data() function need not look at the projected inode.
> 
> Um, the client could still have dirtied data (like a truncate) which requires looking at the projected inode.
>

Won't the dirtied data be flushed back before the scrub uninline work successfully getting the `xlock` ? After that no clients could buffer the new change.